pkg/services/ngalert/schedule/fetcher.go

```diff
@@ -16,10 +18,24 @@ func (sch *schedule) getAlertRules(ctx context.Context, disabledOrgs []int64) []
 	q := models.GetAlertRulesForSchedulingQuery{
 		ExcludeOrgIDs: disabledOrgs,
 	}
+
 	err := sch.ruleStore.GetAlertRulesForScheduling(ctx, &q)
 	if err != nil {
-		sch.log.Error("failed to fetch alert definitions", "err", err)
-		return nil
+		sch.log.Error("failed to fetch alert definitions - using cache", "err", err, "cached_rules", len(sch.currentRuleSet))
+		return sch.currentRuleSet
 	}
+
+	sch.CacheRuleSet(q.Result)
 	return q.Result
 }
+
+// CacheRuleSet caches the current rule set in memory to avoid database failures halting alert evaluations.
+func (sch *schedule) CacheRuleSet(result []*models.SchedulableAlertRule) {
+	h := sha256.New()
+	for i := range result {
+		_, _ = h.Write([]byte(fmt.Sprintf("%v", result[i])))
+	}
+
+	sch.currentRuleSet = result
+	sch.currentRuleSetHash = fmt.Sprintf("%x", h.Sum(nil))
```

Comment:
not used here, but hoping that the next PR can make use of this to avoid constant updates.

Reply:
I'm not sure `result` is sorted so the same set of alert rules will produce different hashes.